File: trunk/open-dis/java/edu/nps/moves/dis/ElectronicEmissionBeamData.java
package edu.nps.moves.dis;

import java.util.*;
import java.io.*;
import javax.xml.bind.annotation.*;

/**
 * Description of one electronic emission beam
 *
 * Copyright (c) 2008, MOVES Institute, Naval Postgraduate School. All rights reserved.
 * This work is licensed under the BSD open source license, available at https://www.movesinstitute.org/licenses/bsd.html
 *
 * @author DMcG
 */
public class ElectronicEmissionBeamData extends Object implements Serializable
{
   /** This field shall specify the length of this beams data in 32 bit words */
   protected short  beamDataLength;

   /** This field shall specify a unique emitter database number assigned to differentiate between otherwise similar or identical emitter beams within an emitter system. */
   protected short  beamIDNumber;

   /** This field shall specify a Beam Parameter Index number that shall be used by receiving entities in conjunction with the Emitter Name field to provide a pointer to the stored database parameters required to regenerate the beam.  */
   protected int  beamParameterIndex;

   /** Fundamental parameter data such as frequency range, beam sweep, etc. */
   protected FundamentalParameterData  fundamentalParameterData = new FundamentalParameterData(); 

   /** beam function of a particular beam */
   protected short  beamFunction;

   /** Number of track/jam targets */
   protected short  numberOfTrackJamTargets;

   /** wheher or not the receiving simulation apps can assume all the targets in the scan pattern are being tracked/jammed */
   protected short  highDensityTrackJam;

   /** padding */
   protected short  pad4 = 0;

   /** identify jamming techniques used */
   protected long  jammingModeSequence;

   /** variable length list of track/jam targets */
   protected List trackJamTargets = new ArrayList(); 

/** Constructor */
 public ElectronicEmissionBeamData()
 {
 }

public int getMarshalledSize()
{
   int marshalSize = 0; 

   marshalSize = marshalSize + 1;  // beamDataLength
   marshalSize = marshalSize + 1;  // beamIDNumber
   marshalSize = marshalSize + 2;  // beamParameterIndex
   marshalSize = marshalSize + fundamentalParameterData.getMarshalledSize();  // fundamentalParameterData
   marshalSize = marshalSize + 1;  // beamFunction
   marshalSize = marshalSize + 1;  // numberOfTrackJamTargets
   marshalSize = marshalSize + 1;  // highDensityTrackJam
   marshalSize = marshalSize + 1;  // pad4
   marshalSize = marshalSize + 4;  // jammingModeSequence
   for(int idx=0; idx < trackJamTargets.size(); idx++)
   {
        TrackJamTarget listElement = (TrackJamTarget)trackJamTargets.get(idx);
        marshalSize = marshalSize + listElement.getMarshalledSize();
   }

   return marshalSize;
}


public void setBeamDataLength(short pBeamDataLength)
{ beamDataLength = pBeamDataLength;
}

@XmlAttribute
public short getBeamDataLength()
{ return beamDataLength; 
}

public void setBeamIDNumber(short pBeamIDNumber)
{ beamIDNumber = pBeamIDNumber;
}

@XmlAttribute
public short getBeamIDNumber()
{ return beamIDNumber; 
}

public void setBeamParameterIndex(int pBeamParameterIndex)
{ beamParameterIndex = pBeamParameterIndex;
}

@XmlAttribute
public int getBeamParameterIndex()
{ return beamParameterIndex; 
}

public void setFundamentalParameterData(FundamentalParameterData pFundamentalParameterData)
{ fundamentalParameterData = pFundamentalParameterData;
}

@XmlElement
public FundamentalParameterData getFundamentalParameterData()
{ return fundamentalParameterData; 
}

public void setBeamFunction(short pBeamFunction)
{ beamFunction = pBeamFunction;
}

@XmlAttribute
public short getBeamFunction()
{ return beamFunction; 
}

@XmlAttribute
public short getNumberOfTrackJamTargets()
{ return (short)trackJamTargets.size();
}

/** Note that setting this value will not change the marshalled value. The list whose length this describes is used for that purpose.
 * The getnumberOfTrackJamTargets method will also be based on the actual list length rather than this value. 
 * The method is simply here for java bean completeness.
 */
public void setNumberOfTrackJamTargets(short pNumberOfTrackJamTargets)
{ numberOfTrackJamTargets = pNumberOfTrackJamTargets;
}

public void setHighDensityTrackJam(short pHighDensityTrackJam)
{ highDensityTrackJam = pHighDensityTrackJam;
}

@XmlAttribute
public short getHighDensityTrackJam()
{ return highDensityTrackJam; 
}

public void setPad4(short pPad4)
{ pad4 = pPad4;
}

@XmlAttribute
public short getPad4()
{ return pad4; 
}

public void setJammingModeSequence(long pJammingModeSequence)
{ jammingModeSequence = pJammingModeSequence;
}

@XmlAttribute
public long getJammingModeSequence()
{ return jammingModeSequence; 
}

public void setTrackJamTargets(List pTrackJamTargets)
{ trackJamTargets = pTrackJamTargets;
}

@XmlElementWrapper(name="trackJamTargetsList" )
public List getTrackJamTargets()
{ return trackJamTargets; }


public void marshal(DataOutputStream dos)
{
    try 
    {
       dos.writeByte( (byte)beamDataLength);
       dos.writeByte( (byte)beamIDNumber);
       dos.writeShort( (short)beamParameterIndex);
       fundamentalParameterData.marshal(dos);
       dos.writeByte( (byte)beamFunction);
       dos.writeByte( (byte)trackJamTargets.size());
       dos.writeByte( (byte)highDensityTrackJam);
       dos.writeByte( (byte)pad4);
       dos.writeInt( (int)jammingModeSequence);

       for(int idx = 0; idx < trackJamTargets.size(); idx++)
       {
            TrackJamTarget aTrackJamTarget = (TrackJamTarget)trackJamTargets.get(idx);
            aTrackJamTarget.marshal(dos);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      System.out.println(e);}
    } // end of marshal method

public void unmarshal(DataInputStream dis)
{
    try 
    {
       beamDataLength = (short)dis.readUnsignedByte();
       beamIDNumber = (short)dis.readUnsignedByte();
       beamParameterIndex = (int)dis.readUnsignedShort();
       fundamentalParameterData.unmarshal(dis);
       beamFunction = (short)dis.readUnsignedByte();
       numberOfTrackJamTargets = (short)dis.readUnsignedByte();
       highDensityTrackJam = (short)dis.readUnsignedByte();
       pad4 = (short)dis.readUnsignedByte();
       jammingModeSequence = dis.readInt();
        for(int idx = 0; idx < numberOfTrackJamTargets; idx++)
        {
           TrackJamTarget anX = new TrackJamTarget();
            anX.unmarshal(dis);
            trackJamTargets.add(anX);
        };

    } // end try 
   catch(Exception e)
    { 
      System.out.println(e); 
    }
 } // end of unmarshal method 


 /**
  * The equals method doesn't always work--mostly on on classes that consist only of primitives. Be careful.
  */
 public boolean equals(ElectronicEmissionBeamData rhs)
 {
     boolean ivarsEqual = true;

    if(rhs.getClass() != this.getClass())
        return false;

     if( ! (beamDataLength == rhs.beamDataLength)) ivarsEqual = false;
     if( ! (beamIDNumber == rhs.beamIDNumber)) ivarsEqual = false;
     if( ! (beamParameterIndex == rhs.beamParameterIndex)) ivarsEqual = false;
     if( ! (fundamentalParameterData.equals( rhs.fundamentalParameterData) )) ivarsEqual = false;
     if( ! (beamFunction == rhs.beamFunction)) ivarsEqual = false;
     if( ! (numberOfTrackJamTargets == rhs.numberOfTrackJamTargets)) ivarsEqual = false;
     if( ! (highDensityTrackJam == rhs.highDensityTrackJam)) ivarsEqual = false;
     if( ! (pad4 == rhs.pad4)) ivarsEqual = false;
     if( ! (jammingModeSequence == rhs.jammingModeSequence)) ivarsEqual = false;

     for(int idx = 0; idx < trackJamTargets.size(); idx++)
     {
        TrackJamTarget x = (TrackJamTarget)trackJamTargets.get(idx);
        if( ! ( trackJamTargets.get(idx).equals(rhs.trackJamTargets.get(idx)))) ivarsEqual = false;
     }


    return ivarsEqual;
 }
} // end of class
// Copyright (c) 1995-2009 held by the author(s).  All rights reserved.
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
//  are met:
// 
//  * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
// * Redistributions in binary form must reproduce the above copyright
// notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
// * Neither the names of the Naval Postgraduate School (NPS)
//  Modeling Virtual Environments and Simulation (MOVES) Institute
// (http://www.nps.edu and http://www.MovesInstitute.org)
// nor the names of its contributors may be used to endorse or
//  promote products derived from this software without specific
// prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// AS IS AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

File: trunk/open-dis/Csharp/DIS/ElectronicEmissionBeamData.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Diagnostics;
using System.Xml.Serialization;
using DISnet.DataStreamUtilities;

namespace DIS1998net
{

/**
 * Description of one electronic emission beam
 *
 * Copyright (c) 2008, MOVES Institute, Naval Postgraduate School. All rights reserved.
 * This work is licensed under the BSD open source license, available at https://www.movesinstitute.org/licenses/bsd.html
 *
 * @author DMcG
 * Modified for use with C#:
 * Peter Smith (Naval Air Warfare Center - Training Systems Division)
 */
[Serializable]
[XmlRoot]
[XmlInclude(typeof(FundamentalParameterData))]
[XmlInclude(typeof(TrackJamTarget))]
public class ElectronicEmissionBeamData : Object
{
   /** This field shall specify the length of this beams data in 32 bit words */
   protected byte  _beamDataLength;

   /** This field shall specify a unique emitter database number assigned to differentiate between otherwise similar or identical emitter beams within an emitter system. */
   protected byte  _beamIDNumber;

   /** This field shall specify a Beam Parameter Index number that shall be used by receiving entities in conjunction with the Emitter Name field to provide a pointer to the stored database parameters required to regenerate the beam.  */
   protected ushort  _beamParameterIndex;

   /** Fundamental parameter data such as frequency range, beam sweep, etc. */
   protected FundamentalParameterData  _fundamentalParameterData = new FundamentalParameterData(); 

   /** beam function of a particular beam */
   protected byte  _beamFunction;

   /** Number of track/jam targets */
   protected byte  _numberOfTrackJamTargets;

   /** wheher or not the receiving simulation apps can assume all the targets in the scan pattern are being tracked/jammed */
   protected byte  _highDensityTrackJam;

   /** padding */
   protected byte  _pad4 = 0;

   /** identify jamming techniques used */
   protected uint  _jammingModeSequence;

   /** variable length list of track/jam targets */
   protected List<object> _trackJamTargets = new List<object>(); 

/** Constructor */
 public ElectronicEmissionBeamData()
 {
 }

public int getMarshalledSize()
{
   int marshalSize = 0; 

   marshalSize = marshalSize + 1;  // _beamDataLength
   marshalSize = marshalSize + 1;  // _beamIDNumber
   marshalSize = marshalSize + 2;  // _beamParameterIndex
   marshalSize = marshalSize + _fundamentalParameterData.getMarshalledSize();  // _fundamentalParameterData
   marshalSize = marshalSize + 1;  // _beamFunction
   marshalSize = marshalSize + 1;  // _numberOfTrackJamTargets
   marshalSize = marshalSize + 1;  // _highDensityTrackJam
   marshalSize = marshalSize + 1;  // _pad4
   marshalSize = marshalSize + 4;  // _jammingModeSequence
   for(int idx=0; idx < _trackJamTargets.Count; idx++)
   {
        TrackJamTarget listElement = (TrackJamTarget)_trackJamTargets[idx];
        marshalSize = marshalSize + listElement.getMarshalledSize();
   }

   return marshalSize;
}


public void setBeamDataLength(byte pBeamDataLength)
{ _beamDataLength = pBeamDataLength;
}

[XmlElement(Type= typeof(byte), ElementName="beamDataLength")]
public byte BeamDataLength
{
     get
{
          return _beamDataLength;
}
     set
{
          _beamDataLength = value;
}
}

public void setBeamIDNumber(byte pBeamIDNumber)
{ _beamIDNumber = pBeamIDNumber;
}

[XmlElement(Type= typeof(byte), ElementName="beamIDNumber")]
public byte BeamIDNumber
{
     get
{
          return _beamIDNumber;
}
     set
{
          _beamIDNumber = value;
}
}

public void setBeamParameterIndex(ushort pBeamParameterIndex)
{ _beamParameterIndex = pBeamParameterIndex;
}

[XmlElement(Type= typeof(ushort), ElementName="beamParameterIndex")]
public ushort BeamParameterIndex
{
     get
{
          return _beamParameterIndex;
}
     set
{
          _beamParameterIndex = value;
}
}

public void setFundamentalParameterData(FundamentalParameterData pFundamentalParameterData)
{ _fundamentalParameterData = pFundamentalParameterData;
}

public FundamentalParameterData getFundamentalParameterData()
{ return _fundamentalParameterData; 
}

[XmlElement(Type= typeof(FundamentalParameterData), ElementName="fundamentalParameterData")]
public FundamentalParameterData FundamentalParameterData
{
     get
{
          return _fundamentalParameterData;
}
     set
{
          _fundamentalParameterData = value;
}
}

public void setBeamFunction(byte pBeamFunction)
{ _beamFunction = pBeamFunction;
}

[XmlElement(Type= typeof(byte), ElementName="beamFunction")]
public byte BeamFunction
{
     get
{
          return _beamFunction;
}
     set
{
          _beamFunction = value;
}
}

public void setHighDensityTrackJam(byte pHighDensityTrackJam)
{ _highDensityTrackJam = pHighDensityTrackJam;
}

[XmlElement(Type= typeof(byte), ElementName="highDensityTrackJam")]
public byte HighDensityTrackJam
{
     get
{
          return _highDensityTrackJam;
}
     set
{
          _highDensityTrackJam = value;
}
}

public void setPad4(byte pPad4)
{ _pad4 = pPad4;
}

[XmlElement(Type= typeof(byte), ElementName="pad4")]
public byte Pad4
{
     get
{
          return _pad4;
}
     set
{
          _pad4 = value;
}
}

public void setJammingModeSequence(uint pJammingModeSequence)
{ _jammingModeSequence = pJammingModeSequence;
}

[XmlElement(Type= typeof(uint), ElementName="jammingModeSequence")]
public uint JammingModeSequence
{
     get
{
          return _jammingModeSequence;
}
     set
{
          _jammingModeSequence = value;
}
}

public void setTrackJamTargets(List<object> pTrackJamTargets)
{ _trackJamTargets = pTrackJamTargets;
}

public List<object> getTrackJamTargets()
{ return _trackJamTargets; }

[XmlElement(ElementName = "trackJamTargetsList",Type = typeof(List<object>))]
public List<object> TrackJamTargets
{
     get
{
          return _trackJamTargets;
}
     set
{
          _trackJamTargets = value;
}
}


public void marshal(DataOutputStream dos)
{
    try 
    {
       dos.writeByte( (byte)_beamDataLength);
       dos.writeByte( (byte)_beamIDNumber);
       dos.writeUshort( (ushort)_beamParameterIndex);
       _fundamentalParameterData.marshal(dos);
       dos.writeByte( (byte)_beamFunction);
       dos.writeByte( (byte)_trackJamTargets.Count);
       dos.writeByte( (byte)_highDensityTrackJam);
       dos.writeByte( (byte)_pad4);
       dos.writeUint( (uint)_jammingModeSequence);

       for(int idx = 0; idx < _trackJamTargets.Count; idx++)
       {
            TrackJamTarget aTrackJamTarget = (TrackJamTarget)_trackJamTargets[idx];
            aTrackJamTarget.marshal(dos);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      Trace.WriteLine(e);
      Trace.Flush();
    }
} // end of marshal method

public void unmarshal(DataInputStream dis)
{
    try 
    {
       _beamDataLength = dis.readByte();
       _beamIDNumber = dis.readByte();
       _beamParameterIndex = dis.readUshort();
       _fundamentalParameterData.unmarshal(dis);
       _beamFunction = dis.readByte();
       _numberOfTrackJamTargets = dis.readByte();
       _highDensityTrackJam = dis.readByte();
       _pad4 = dis.readByte();
       _jammingModeSequence = dis.readUint();
        for(int idx = 0; idx < _numberOfTrackJamTargets; idx++)
        {
           TrackJamTarget anX = new TrackJamTarget();
            anX.unmarshal(dis);
            _trackJamTargets.Add(anX);
        };

    } // end try 
   catch(Exception e)
    { 
      Trace.WriteLine(e); 
      Trace.Flush();
    }
 } // end of unmarshal method 


public void reflection(StringBuilder sb)
{
    sb.Append("----- ElectronicEmissionBeamData-----"  + System.Environment.NewLine);
    try 
    {
           sb.Append("byte\t _beamDataLength\t " + _beamDataLength.ToString() + System.Environment.NewLine);
           sb.Append("byte\t _beamIDNumber\t " + _beamIDNumber.ToString() + System.Environment.NewLine);
           sb.Append("ushort\t _beamParameterIndex\t " + _beamParameterIndex.ToString() + System.Environment.NewLine);
       sb.Append("=====_fundamentalParameterData=====" + System.Environment.NewLine);
       _fundamentalParameterData.reflection(sb);
           sb.Append("byte\t _beamFunction\t " + _beamFunction.ToString() + System.Environment.NewLine);
           sb.Append("byte\t _trackJamTargets\t " + _trackJamTargets.Count.ToString() + System.Environment.NewLine);
           sb.Append("byte\t _highDensityTrackJam\t " + _highDensityTrackJam.ToString() + System.Environment.NewLine);
           sb.Append("byte\t _pad4\t " + _pad4.ToString() + System.Environment.NewLine);
           sb.Append("uint\t _jammingModeSequence\t " + _jammingModeSequence.ToString() + System.Environment.NewLine);

       for(int idx = 0; idx < _trackJamTargets.Count; idx++)
       {
           sb.Append("TrackJamTarget\t " + _trackJamTargets[idx] + System.Environment.NewLine);
            TrackJamTarget aTrackJamTarget = (TrackJamTarget)_trackJamTargets[idx];
            aTrackJamTarget.reflection(sb);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      Trace.WriteLine(e);
      Trace.Flush();
}
    } // end of marshal method

 /**
  * The equals method doesn't always work--mostly it works only on on classes that consist only of primitives. Be careful.
  */
 public bool equals(ElectronicEmissionBeamData rhs)
 {
     bool ivarsEqual = true;

    if(rhs.GetType() != this.GetType())
        return false;

     if( ! (_beamDataLength == rhs._beamDataLength)) ivarsEqual = false;
     if( ! (_beamIDNumber == rhs._beamIDNumber)) ivarsEqual = false;
     if( ! (_beamParameterIndex == rhs._beamParameterIndex)) ivarsEqual = false;
     if( ! (_fundamentalParameterData.Equals( rhs._fundamentalParameterData) )) ivarsEqual = false;
     if( ! (_beamFunction == rhs._beamFunction)) ivarsEqual = false;
     if( ! (_numberOfTrackJamTargets == rhs._numberOfTrackJamTargets)) ivarsEqual = false;
     if( ! (_highDensityTrackJam == rhs._highDensityTrackJam)) ivarsEqual = false;
     if( ! (_pad4 == rhs._pad4)) ivarsEqual = false;
     if( ! (_jammingModeSequence == rhs._jammingModeSequence)) ivarsEqual = false;

     for(int idx = 0; idx < _trackJamTargets.Count; idx++)
     {
        TrackJamTarget x = (TrackJamTarget)_trackJamTargets[idx];
        if( ! ( _trackJamTargets[idx].Equals(rhs._trackJamTargets[idx]))) ivarsEqual = false;
     }


    return ivarsEqual;
 }
} // end of class
} // end of namespace
